include/beman/inplace_vector/inplace_vector.hpp

```diff
@@ -0,0 +1,1034 @@
+    __IV_EXPECT(__i < __N);
+    return reinterpret_cast<__T *>(__d) + __i;
+  }
+  constexpr const __T *__data(size_t __i) const noexcept {
+    __IV_EXPECT(__i < __N);
+    return reinterpret_cast<const __T *>(__d) + __i;
+  }
+};
+
+// Storage for zero elements.
+template <class __T> struct __zero_sized {
+protected:
+  using __size_type = uint8_t;
+  static constexpr __T *__data() noexcept { return nullptr; }
+  static constexpr __size_type __size() noexcept { return 0; }
+  static constexpr void __unsafe_set_size(size_t __new_size) noexcept {
+    __IV_EXPECT(__new_size == 0 &&
+                "tried to change size of empty storage to non-zero value");
+  }
+
+public:
+  constexpr __zero_sized() = default;
+  constexpr __zero_sized(__zero_sized const &) = default;
+  constexpr __zero_sized &operator=(__zero_sized const &) = default;
+  constexpr __zero_sized(__zero_sized &&) = default;
+  constexpr __zero_sized &operator=(__zero_sized &&) = default;
+  constexpr ~__zero_sized() = default;
+};
+
+// Storage for trivial types.
+template <class __T, size_t __N> struct __trivial {
+  static_assert(is_trivial_v<__T>,
+                "storage::trivial<T, C> requires Trivial<T>");
+  static_assert(__N != size_t{0}, "__N  == 0, use __zero_sized");
+
+protected:
+  using __size_type = __smallest_size_t<__N>;
+
+private:
+  // If value_type is const, then const array of non-const elements:
+  using __data_t = conditional_t<!is_const_v<__T>, array<__T, __N>,
+                                 const array<remove_const_t<__T>, __N>>;
+  alignas(alignof(__T)) __data_t __data_{};
+  __size_type __size_ = 0;
+
+protected:
+  constexpr const __T *__data() const noexcept { return __data_.data(); }
+  constexpr __T *__data() noexcept { return __data_.data(); }
+  constexpr __size_type __size() const noexcept { return __size_; }
+  constexpr void __unsafe_set_size(size_t __new_size) noexcept {
+    __IV_EXPECT(__size_type(__new_size) <= __N &&
+                "new_size out-of-bounds [0, N]");
+    __size_ = __size_type(__new_size);
+  }
+
+public:
+  constexpr __trivial() noexcept = default;
+  constexpr __trivial(__trivial const &) noexcept = default;
+  constexpr __trivial &operator=(__trivial const &) noexcept = default;
+  constexpr __trivial(__trivial &&) noexcept = default;
+  constexpr __trivial &operator=(__trivial &&) noexcept = default;
+  constexpr ~__trivial() = default;
+};
+
+/// Storage for non-trivial elements.
+template <class __T, size_t __N> struct __non_trivial {
+  static_assert(!is_trivial_v<__T>,
+                "use storage::trivial for Trivial<T> elements");
+  static_assert(__N != size_t{0}, "use storage::zero for __N==0");
+
+protected:
+  using __size_type = __smallest_size_t<__N>;
+
+private:
+  using __data_t =
+      conditional_t<!is_const_v<__T>, __aligned_storage2<__T, __N>,
+                    const __aligned_storage2<remove_const_t<__T>, __N>>;
+  __data_t __data_{}; // BUGBUG: test SIMD types
+  __size_type __size_ = 0;
```

Comment:
we could use a union based lifetime delegation here instead of byte array based type erasure for supporting trivial union based constexpr:
```
union { __T __data_[__N]; };
__size_type __size_ = 0;
```

This should cover both non-trivial base case and trivial union based constexpr implementation.

Reply:
I think there is no point switch to a union based approach before the issues addressed in [P3074](https://wg21.link/p3074 )are fixed. As of now, we would need to add a constructor and destructor to the union (as you write above), therefore the storage (and thus the whole inplace_vector) could no longer be trivially constructible and trivially destructible.
Note that the requirement

> If is_trivially_destructible_v<T> is true, then:
> IV has a trivial destructor

holds for any T, independent of its triviality (and thus, as of now, storage type).

Reply:
A general notes on the tests:

- Tests should achieve a high (ideally full) coverage of all the branches, this includes not only runtime branches, but also "implementation branches" due to, e.g., different storage types for trivial vs non-trivial types.
- Ideally, we make the tests independent from the implementation (black box) and only derive the test cases from the specification.
- The current specification tells us that there must be at least two implementation branches for trivial and non-trivial T, so we should at least test with such types
- The requirements on special member functions provide more information, for example

  >  If is_trivially_destructible_v is true, then:
  >    IV has a trivial destructor

  this means that we need to test the destructor (at least) with 1. a trivial type, 2. a non-trivial type which is trivially destructible, 3. a non-trivial type which is not trivially destructible
- Further requirements on other other special member functions make it even more complicated
- In addition there are some special requirements (and thus an implementation branch) for size 0.

All in all, we need to determine the proper combination of types and then run the full test suite on them.
Once we have that we can fix the implementation and experiment with improvements.
(We might even be able to port back the implementation to C++20, ...)

I have some WIP with this respect which I hope I can commit shortly.

Reply:
> we would need to add a constructor and destructor to the union (as you write above), therefore the storage (and thus the whole inplace_vector) could no longer be trivially constructible and trivially destructible.

You are right, I missed that.

I support backporting to C++ 20, the original aim was C++20 for trivial constexpr, C++17 for basic functionality. 

We might need to also test for const T.